Hotfix/#34 removetag (#34)

jiro4989 opened the pull request:
#34 @theoldmoon0602 
テストコードを書いてみたら別に不具合はなさそうでしたので
テストコードのプルリクエストとなります。
テストコードを書くためにテストコードライブラリをGopkg.tomlに追加しています。

jiro4989 committed:
:white_check_mark: タグを削除する処理を関数に分離してテストコードを追加

jiro4989 committed:
:heavy_minus_sign: anacondaのバージョンを上げてしまっていたので戻した

theoremoon commented on tweet.go:
```diff
@@ -96,6 +87,20 @@ func ExtractShellgei(tweet anaconda.Tweet, self anaconda.User, api *anaconda.Twi
+	for _, tag := range entities.Hashtags {
+		for _, t := range tags {
+			if tag.Text == t {
+				rtext = append(rtext[:tag.Indices[0]-deletecount], rtext[tag.Indices[1]-deletecount:]...)
```
sliceをよく理解しないままこれを書いてしまったんですが、ここが原因でpanicしてしまうケースがあるようです。メモリのアクセス違反などを疑っていますがそういう可能性はありませんか？　（その場合どう書くのが良いのかちょっとわからないんですが）

以下エラー例です
```
panic: runtime error: slice bounds out of range

goroutine 86 [running]:
main.ExtractShellgei(0x0, 0x0, 0x0, 0x0, 0xc00025a1a0, 0x1e, 0x0, 0x0, 0x0, 0xc000265400, ...)
	/go/src/github.com/theoldmoon0602/ShellgeiBot/tweet.go:74 +0x13bf
main.ProcessTweet(0x0, 0x0, 0x0, 0x0, 0xc00025a1a0, 0x1e, 0x0, 0x0, 0x0, 0xc000265400, ...)
	/go/src/github.com/theoldmoon0602/ShellgeiBot/main.go:44 +0x269
main.botMain.func1(0xc00011bc00, 0xc00009c780, 0xc000096230, 0xc0000de0c0, 0xc0000ee1e0)
	/go/src/github.com/theoldmoon0602/ShellgeiBot/main.go:115 +0x162
created by main.botMain
	/go/src/github.com/theoldmoon0602/ShellgeiBot/main.go:114 +0x9f7
```

theoremoon reviewed (changes requested):
コメントしたとおりおそらくバグが残っているので修正お願いします（自分でかけないのに偉そうでごめんなさい :bow:

jiro4989 commented:
なるほどなるほど。承知しました。
エラーが発生したときにどういうデータを送ったのかも教えてくださいますと助かります :pray:

theoremoon commented:
@jiro4989 このデータの場合はおそらく[このツイート](https://twitter.com/air_murasaki/status/1151449231187595265)に対してのものですが他にも数回確認しているので（それはどのツイートに対してかわかっていません。申し訳ないです）、おそらくデータの中身とはあまり関係ないと思っています

jiro4989 commented:
@theoldmoon0602 

こんな感じのデータをテストコードに追加したら同じようなエラーがでました。

```go
			text:   "echo test #シェル芸 #シェル芸 #シェル芸",
```

```
--- FAIL: TestRemoveTags (0.00s)
panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 7 [running]:
testing.tRunner.func1(0xc0000f8200)
	/usr/local/src/go1.12/go/src/testing/testing.go:830 +0x388
panic(0x839820, 0xd98f80)
	/usr/local/src/go1.12/go/src/runtime/panic.go:522 +0x1b5
github.com/jiro4989/ShellGeiBot.removeTags(0x8ab644, 0x33, 0xc0000f6200, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/jiro4989/src/github.com/jiro4989/ShellGeiBot/tweet.go:96 +0x413
github.com/jiro4989/ShellGeiBot.TestRemoveTags(0xc0000f8200)
	/home/jiro4989/src/github.com/jiro4989/ShellGeiBot/tweet_test.go:228 +0x81c
testing.tRunner(0xc0000f8200, 0x8b2d50)
	/usr/local/src/go1.12/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/src/go1.12/go/src/testing/testing.go:916 +0x357
exit status 2
FAIL	github.com/jiro4989/ShellGeiBot	0.005s
```

スライス操作についてじっくり見てみますね

jiro4989 committed:
:bug: エラーが発生するテストデータを追加 (まだ直していない)

jiro4989 commented:
あ、ごめんなさい。テストデータが間違ってました

jiro4989 committed:
:white_check_mark: テストデータが間違っていたので修正

jiro4989 commented:
そうですねー。同じシェルを実行してみても今度は処理が完了しましたし、
データは直接関係なさそうですね
https://twitter.com/jiro_saburomaru/status/1151457539113885697

theoremoon commented:
あーわかったかもしれない。今のコードだと前から順番にタグが削除されているかのように書いてるけど、後ろのタグが先に削除されるとインデックスがずれるんですかね

theoremoon commented:
いやあでもそれだとタグ1個のときのエラーが説明できませんか

jiro4989 commented:
@theoldmoon0602 tweet.Entities.Hashtagsが返すタグの順序が、タグが出現する位置の数値の小さい順に返すとは限らない場合、仰る現象は起きそうですね :thinking:

theoremoon commented:
hashtagの順序についてドキュメントには書いてないですが、サンプルだと一応出現順になっているのでこれは大丈夫だと思っても良いんですかね。それだと原因やっぱりわからないですが

jiro4989 commented:
@theoldmoon0602 ああ、ごめんなさい。作業ブランチをmasterから切ってしまってました。
https://github.com/theoldmoon0602/ShellgeiBot/blob/8272fbbdd9b6897652c074dbb8ff5932459cd662/tweet.go#L68
こちらでは、 Prefix を取り除いて判定するようになっていましたね。

theoremoon commented:
@jiro4989 あーいや、それは古いやつですね（確かにわざわざ固定リンクを貼ったのでわかりにくくしてしまったかも）。TrimPrefixは不要と判断したので削除しました（実際入力データに#はないので意味がないです

jiro4989 commented:
なるほど、そうでしたか。承知しました。

theoremoon commented on tweet.go:
```diff
@@ -66,17 +67,7 @@ func ExtractShellgei(tweet anaconda.Tweet, self anaconda.User, api *anaconda.Twi
-		}
-	}
-	shellgei := strings.TrimSpace(string(rtext))
+	shellgei := removeTags(text, tweet.Entities, tags)
```
ツイートが140文字以上の場合ハッシュタグは ExtendedEntitiesに格納されるらしいので tweet.ExtendedEntitiesに対しても同じ処理が必要になりそうですね（おそらくどちらかに集中してタグが格納されるのでとりあえず両方やるので良さそうです）

jiro4989 replied:
そうですね、公式APIリファレンスにはそのように記載がありました。
気になったので手元で別にannacondaのコードを書いて試してみたところ、Extended Entitiesは空みたいでした :thinking: 

とはいえ、リファレンスには記載ありますし対応したほうが良さそうですね。

```
text: echo test #シェル芸 #シェル芸1 #シェル芸2 #シェル芸3 #シェル芸4 #シェル芸5 #シェル芸6 #シェル芸7 #シェル芸8 #シェル芸9 #シェル芸10 #シェル芸11 #シェル芸12 #シェル芸13 #シェル芸14 #シェル芸15 #シェル芸16 #シェル芸17 #シェル芸18 #シェル芸19 #シェル芸20
tag: [{[10 15] シェル芸} {[16 22] シェル芸1} {[23 29] シェル芸2} {[30 36] シェル芸3} {[37 43] シェル芸4} {[44 50] シェル芸5} {[51 57] シェル芸6} {[58 64] シェル芸7} {[65 71] シェル芸8} {[72 78] シェル芸9} {[79 86] シェル芸10} {[87 94] シェル芸11} {[95 102] シェル芸12} {[103 110] シェル芸13} {[111 118] シェル芸14} {[119 126] シェル芸15} {[127 134] シェル芸16} {[135 142] シェル芸17} {[143 150] シェル芸18} {[151 158] シェル芸19} {[159 166] シェル芸20}]
ext tag: []

text: echo 12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890
#シェル芸
tag: [{[206 211] シェル芸}]
ext tag: []
```


```go
package main

import (
	"fmt"
	"net/url"

	"github.com/ChimeraCoder/anaconda"
)

func getTwitterApi() *anaconda.TwitterApi {
	anaconda.SetConsumerKey("xxx")
	anaconda.SetConsumerSecret("xxx")
	return anaconda.NewTwitterApi("xxx", "xxx")
}

func main() {
	api := getTwitterApi()

	v := url.Values{}
	v.Set("count", "10")

	searchResult, _ := api.GetSearch("#シェル芸", v)
	for _, tweet := range searchResult.Statuses {
		fmt.Println("text:", tweet.FullText)
		fmt.Println("tag:", tweet.Entities.Hashtags)
		fmt.Println("ext tag:", tweet.ExtendedEntities.Hashtags)
		fmt.Println("---------")
	}
}

```

ikawaha commented:
完全に横からですみません。
効率はよくないかもですが、マークしてから消し込む感じで処理するのでどうでしょうか。
（strings.TrimSpace() してから返却でもいいかもしれない・・・）

```go
type hashTags []struct{
	Indices []int
	Text    string
}
func removeTags(s string, tags, exTags hashTags) string {
	text := []rune(s)
	for _, tags := range []hashTags{tags, exTags}{
		for _, v :=range tags {
			if len(v.Indices) < 2 {
				continue // or error or panic ???
			}
			for i := v.Indices[0]; i < v.Indices[1]; i++ {
				text[i] = rune(0)
			}
		}
	}
	var b bytes.Buffer
	for _, v :=range text {
		if v != rune(0) {
			b.WriteRune(v)
		}
	}
	return b.String()
}
```

呼び出しは
```go
	text = removeTags(text, tweet.Entities.Hashtags, tweet.ExtendedEntities.Hashtags)
```
こんな想定です。


テストはこんな感じで事例を追加していけるかなと思います。
```go
func TestRemoveTags(t *testing.T) {
	cases :=[]struct{
		Name string
		Text string
		Tags hashTags
		ExtendedTags hashTags
		Expected string
	}{
		{
			Name: "タグが20個, Exはなし",
			Text: "echo test #シェル芸 #シェル芸1 #シェル芸2 #シェル芸3 #シェル芸4 #シェル芸5 #シェル芸6 #シェル芸7 #シェル芸8 #シェル芸9 #シェル芸10 #シェル芸11 #シェル芸12 #シェル芸13 #シェル芸14 #シェル芸15 #シェル芸16 #シェル芸17 #シェル芸18 #シェル芸19 #シェル芸20",
			Tags: hashTags{
				{Indices: []int{10,15}, Text: "シェル芸"},
				{Indices: []int{16, 22}, Text: "シェル芸1"},
				{Indices: []int{23, 29}, Text: "シェル芸2"},
				{Indices: []int{30, 36}, Text: "シェル芸3"},
				{Indices: []int{37, 43}, Text: "シェル芸4"},
				{Indices: []int{44, 50}, Text: "シェル芸5"},
				{Indices: []int{51, 57}, Text: "シェル芸6"},
				{Indices: []int{58, 64}, Text: "シェル芸7"},
				{Indices: []int{65, 71}, Text: "シェル芸8"},
				{Indices: []int{72, 78}, Text: "シェル芸9"},
				{Indices: []int{79, 86}, Text: "シェル芸10"},
				{Indices: []int{87, 94}, Text: "シェル芸11"},
				{Indices: []int{95, 102}, Text: "シェル芸12"},
				{Indices: []int{103, 110}, Text: "シェル芸13"},
				{Indices: []int{111, 118}, Text: "シェル芸14"},
				{Indices: []int{119, 126}, Text: "シェル芸15"},
				{Indices: []int{127, 134}, Text: "シェル芸16"},
				{Indices: []int{135, 142}, Text: "シェル芸17"},
				{Indices: []int{143, 150}, Text: "シェル芸18"},
				{Indices: []int{151, 158}, Text: "シェル芸19"},
				{Indices: []int{159, 166}, Text: "シェル芸20"},
			},
			ExtendedTags: hashTags{},
			Expected: "echo test",
		},
	}
	for _, c := range cases {
		t.Run(c.Name, func(t *testing.T) {
			if got :=removeTags(c.Text, c.Tags, c.ExtendedTags); strings.TrimSpace(got) !=c.Expected{
				t.Errorf("expected '%v', got '%v'", c.Expected, got)
			}
		})
	}
}
```

jiro4989 commented:
@ikawaha マークをつけて削除するというアプローチ、とてもスマートでわかりやすく思います。

一つ懸念があるとするならば、提案いただいたアプローチの場合、もともとのテキストに含まれるNULL文字も削除される可能性がある、ということでしょうか。

シェルのコマンドの実行結果にNULL文字が含まれることはよくありますが、
つぶやき自体（シェルのコマンド自体）にNULL文字が含まれるケースは、意図的にそうしない限りレアケースである、と思いますので気にしすぎかもしれませんが。

ikawaha commented:
ふわー。NULL文字許容なのですか。なめてました。奥深い。
unicode の非文字 U+FFFE とかを使う必要がありそうですね😇

theoremoon commented:
@ikawaha あーよさそう。ありがとうございます。特に遅そうな感じもしないのでこの方針採用したいですね

jiro4989 committed:
:recycle: Tweet.Entities.Hashtagsを切り出した構造体定義の追加とそれを使用するように変更

これによりテストコードの記述を容易にする.

https://github.com/theoldmoon0602/ShellgeiBot/pull/35#issuecomment-512612492

jiro4989 committed:
:recycle: マークをつけてから削除する方針に変更

jiro4989 committed:
:recycle: 削除マークをUNICODE非文字に変更

jiro4989 committed:
:recycle: Extended entites hashtagsも精査するように変更

jiro4989 committed:
:white_check_mark: タグが20個のケースを追加

theoremoon committed:
:pill: 構造体の型を修正

theoremoon committed:
:heart: 好みに合わせて変更

theoremoon reviewed (approved):
これでいきます

theoremoon commented:
どーん
